types/react/index.d.ts

```diff
@@ -37,33 +52,128 @@ declare namespace React {
     // React Elements
     // ----------------------------------------------------------------------
 
+    /**
+     * Used to retrieve the possible components which accept a given set of props.
+     *
+     * Can be passed no type parameters to get a union of all possible components
+     * and tags.
+     *
+     * Is a superset of {@link ComponentType}.
+     *
+     * @template P The props to match against. If not passed, defaults to any.
+     * @template Tag An optional tag to match against. If not passed, attempts to match against all possible tags.
+     *
+     * @example
+     *
+     * ```tsx
+     * // All components and tags (img, embed etc.)
+     * // which accept `src`
+     * type SrcComponents = ElementType<{ src: any }>;
+     * ```
+     *
+     * @example
+     *
+     * ```tsx
+     * // All components
+     * type AllComponents = ElementType;
+     * ```
+     *
+     * @example
+     *
+     * ```tsx
+     * // All custom components which match `src`, and tags which
+     * // match `src`, narrowed down to just `audio` and `embed`
+     * type SrcComponents = ElementType<{ src: any }, 'audio' | 'embed'>;
+     * ```
+     */
     type ElementType<P = any, Tag extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements> =
         | { [K in Tag]: P extends JSX.IntrinsicElements[K] ? K : never }[Tag]
         | ComponentType<P>;
```

Comment:
Hey, I'm not convinced by saying that it is used to "retrieve the possible components", since `"img"`, for example, is not a component, but a "tag name". I just found that [here](https://html.spec.whatwg.org/multipage/syntax.html#elements-2).


That being said, in the example below `type AllComponents` wouldn't be right.

```suggestion
    /**
     * Used to retrieve the possible tag names and components which accept a given set of props.
     *
     * Can be passed no type parameters to get a union of all possible tag names and components
     *
     * Is a superset of {@link ComponentType}.
     *
     * @template P The props to match against. If not passed, defaults to any.
     * @template Tag An optional tag name to match against. If not passed, attempts to match against all possible elements.
     *
     * @example
     *
     * ```tsx
     * // All components and tag names (img, embed etc.)
     * // which accept `src`
     * type SrcComponents = ElementType<{ src: any }>;
     * ```
     *
     * @example
     *
     * ```tsx
     * // All components
     * type AllComponents = ElementType;
     * ```
     *
     * @example
     *
     * ```tsx
     * // All custom components which match `src`, and tag names which
     * // match `src`, narrowed down to just `audio` and `embed`
     * type SrcComponents = ElementType<{ src: any }, 'audio' | 'embed'>;
     * ```
     */
    type ElementType<P = any, Tag extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements> =
        | { [K in Tag]: P extends JSX.IntrinsicElements[K] ? K : never }[Tag]
        | ComponentType<P>;
```

What do you think?

Reply:
If we take `<img />` as JSX then `'img'` is the component not a tag name. We're using the same terminology in the official docs e.g. https://react.dev/reference/react-dom/components/input